### ISSUE TITLE
How to get pcode in the right order

### ISSUE THREAD

**Issue by RedRNG** [NationalSecurityAgency/ghidra#2023]
I am currently trying to make a script to improve the auto analysis.

I encountered 2 problems during the process.

### First Problem : getting pcode in execution order
The first is that I can't manage to get the Pcode instructions in execution order.

My script :
```
Iterator<PcodeOpAST> pcodeAst = hf.getPcodeOps();
		
while (pcodeAst.hasNext()) {
		
	PcodeOpAST pcodeast = pcodeAst.next();

	PrintAST(pcodeast);
}
```

For example, with the following C instruction :
`*(undefined *)((longlong)param_1 + 0xae) = 0x14;`

The resulting output :

> PTRSUB : (unique, 0x100000ab, 8) = (unique, 0x100000a3, 8) + (const, 0xae, 8)
> STORE : *(unique, 0x640, 8) = (const, 0x14, 1)
> CAST : (unique, 0x100000a3, 8) = (longlong) param_1
> CAST : (unique, 0x640, 8) = (undefined\*) (unique, 0x100000ab, 8)

The expected output (in execution order) : 

> CAST : (unique, 0x100000a3, 8) = (longlong) param_1
> PTRSUB : (unique, 0x100000ab, 8) = (unique, 0x100000a3, 8) + (const, 0xae, 8)
> CAST : (unique, 0x640, 8) = (undefined\*) (unique, 0x100000ab, 8)
> STORE : *(unique, 0x640, 8) = (const, 0x14, 1)

### Second problem : retrieving the name of the variables
The second one is that I am trying to follow up the tree from a specific opcode to a named variable (local or param) so that I can set a new datatype to that variable for example.

My script :
```
HighVariable getNamedVariable(Varnode vnode) {
     HighVariable hvar = vnode.getHigh();
     if (hvar.getName() != null)
       return hvar;

     // Calling back getNamedVariable() on all of the pcode's input nodes
}
```
This script works for most cases but on certain instruction I end up on a PTRSUB of a register + a constant which have a null opcode.
Even though I know for sure the code is related to a named variable.

For example with the following C instructions:
```
string_2 *local_88;
list_string_2_destructor((std_list_string_2 *)&local_88);
```

The resulting output :

>CALL : [(ram, 0x14023c040, 8)], std_list_string_2 * (unique, 0x620, 8)
>looking through : (unique, 0x620, 8)
>CAST : (unique, 0x620, 8) = (std_list_string_2 *) (unique, 0x100006e8, 8)
>looking through : (unique, 0x100006e8, 8)
>PTRSUB : temporary_(unique, 0x100006e8, 8) = (register, 0x20, 8) + (const, 0xffffffffffffff78, 8)
>looking through : (register, 0x20, 8)
>Nothing.
>looking through : (const, 0xffffffffffffff78, 8)
>Nothing.

The expected result :

>CALL : [(ram, 0x14023c040, 8)], std_list_string_2 * (unique, 0x620, 8)
>looking through : (unique, 0x620, 8)
>CAST : (unique, 0x620, 8) = (std_list_string_2 *) (unique, 0x100006e8, 8)
>ooking through : (unique, 0x100006e8, 8)
>PTRSUB : temporary_(unique, 0x100006e8, 8) = (register, 0x20, 8) + (const, 0xffffffffffffff78, 8)
>looking through : (register, 0x20, 8)
>....
> Found reference to variable local_88

### Specs

OS : Windows 10 x64
Ghidra : 9.1.2

**Comment by astrelsky** [comment]
> ### Second problem : retrieving the name of the variables
> The second one is that I am trying to follow up the tree from a specific opcode to a named variable (local or param) so that I can set a new datatype to that variable for example.
> 
> My script :
> 
> ```
> HighVariable getNamedVariable(Varnode vnode) {
>      HighVariable hvar = vnode.getHigh();
>      if (hvar.getName() != null)
>        return hvar;
> 
>      // Calling back getNamedVariable() on all of the pcode's input nodes
> }
> ```

Just a heads up, if that `Varnode` did **not** come from `DecompileResults` then you have a raw varnode. The varnodes you get from the listing are always raw varnode's. This is what the javadocs says "Rawest possible Varnode. Just a variable location and size, not part of a syntax tree. A raw varnode is said to be free, it is not attached to any variable." ie it's useless for analysis.

**Comment by RedRNG** [comment]
Yes I forgot to say that every Varnode come from the DecompileResults.

Here is the main function:
```
void ImproveFunctionResult(Function f, DecompInterface dec) throws Exception {
        DecompileResults res = dec.decompileFunction(f, 0, monitor);
		   
	if (!res.decompileCompleted()) {
           println(res.getErrorMessage());
	   return;
	}
		
        ImproveDecompiler(res);
}
```

**Comment by Dittozzz** [comment]
Have you solved the first problem? I met the same problem

**Comment by bdemick** [comment]
I do think there may be a bug in the `PcodeOpAST` iterator implementation or at least it isn't implemented in the most straightforward way. One would expect them to be in execution order but they are definitely not. My workaround for this has been to display by blocks (Python pseudocode):

```
high_func = ifc.decompileFunction(func, 60, TaskMonitor.DUMMY).getHighFunction()

for block in high_func.getBasicBlocks():
    print "\nloc_" + str(block.start)
    itr = block.getIterator()
    while itr.hasNext():
        print itr.next()
````

Which seems to always display in the proper order, with the added bonus of having better visualization of blocks.

**Comment by jamesguo71** [comment]
@bdemick
But it seems that your workaround doesn't solve the problem. I tried it but apparent the "loc_Addr" are not ordered. I am thinking about sorting the blocks based on the block.start to make it work.

**Comment by bdemick** [comment]
The issue is that the `PcodeOpAST` iterator is a bit buggy, but the block iterator isn't, for some reason. E.g., when traversing the PCodeOps in a block, you get them in the expected order. And of course, you'll have to order the blocks (or better yet, embed them in a graph view). It makes more sense to view them as blocks, anyway, IMO.
